core/components/tree/basenode.js

```diff
@@ -0,0 +1,1151 @@
+        }
+      }
+    }
+  }
+};
+
+
+/**
+ * Adds a node as a child to the current node.
+ * @param {Blockly.tree.BaseNode} child The child to add.
+ * @param {Blockly.tree.BaseNode=} opt_before If specified, the new child is
+ *    added as a child before this one. If not specified, it's appended to the
+ *    end.
+ * @return {!Blockly.tree.BaseNode} The added child.
+ */
+Blockly.tree.BaseNode.prototype.add = function(child, opt_before) {
+  if (child.getParent()) {
+    child.getParent().removeChild(child);
+  }
+  this.addChildAt(
+      child, opt_before ? this.indexOfChild(opt_before) : this.getChildCount());
+  return child;
+};
+
+/**
+ * Returns the tree.
+ * @return {?Blockly.tree.TreeControl} tree
+ */
+Blockly.tree.BaseNode.prototype.getTree = function() {
+  return null;
+};
+
+/**
+ * Returns the depth of the node in the tree. Should not be overridden.
+ * @return {number} The non-negative depth of this node (the root is zero).
+ */
+Blockly.tree.BaseNode.prototype.getDepth = function() {
+  var depth = this.depth_;
+  if (depth < 0) {
+    depth = this.computeDepth_();
+    this.setDepth_(depth);
+  }
+  return depth;
+};
+
+/**
+ * Computes the depth of the node in the tree.
+ * Called only by getDepth, when the depth hasn't already been cached.
+ * @return {number} The non-negative depth of this node (the root is zero).
+ * @private
+ */
+Blockly.tree.BaseNode.prototype.computeDepth_ = function() {
+  var parent = this.getParent();
+  if (parent) {
+    return parent.getDepth() + 1;
+  } else {
+    return 0;
+  }
+};
+
+/**
+ * Changes the depth of a node (and all its descendants).
+ * @param {number} depth The new nesting depth; must be non-negative.
+ * @private
+ */
+Blockly.tree.BaseNode.prototype.setDepth_ = function(depth) {
+  if (depth != this.depth_) {
+    this.depth_ = depth;
+    var row = this.getRowElement();
+    if (row) {
+      var indent = this.getPixelIndent_() + 'px';
+      if (this.isRightToLeft()) {
+        row.style.paddingRight = indent;
+      } else {
+        row.style.paddingLeft = indent;
+      }
+    }
+    this.forEachChild(function(child) { child.setDepth_(depth + 1); });
+  }
+};
+
+/**
+ * Returns true if the node is a descendant of this node
+ * @param {Blockly.tree.BaseNode} node The node to check.
+ * @return {boolean} True if the node is a descendant of this node, false
+ *    otherwise.
+ */
+Blockly.tree.BaseNode.prototype.contains = function(node) {
+  var current = node;
+  while (current) {
+    if (current == this) {
+      return true;
+    }
+    current = current.getParent();
+  }
+  return false;
+};
+
+/**
+ * An array of empty children to return for nodes that have no children.
+ * @type {!Array<!Blockly.tree.BaseNode>}
+ * @private
+ */
+Blockly.tree.BaseNode.EMPTY_CHILDREN_ = [];
+
+/**
+ * @param {number} index 0-based index.
+ * @return {Blockly.tree.BaseNode} The child at the given index; null if none.
+ */
+Blockly.tree.BaseNode.prototype.getChildAt;
```

Comment:
???

Reply:
There's no need for that, I don't know why closure does that, will remove.

Reply:
Component.getChildAt sets it's return type as Blockly.Component. This is just to please the closure compiler so it doesn't throw warnings.